Stripe/STPRedirectClient.m

```diff
@@ -0,0 +1,157 @@
+
+    _inProgressAuthSource = source;
+
+    [[STPURLCallbackHandler shared] registerListener:self
+                                              forURL:returnURL];
+
+    if (!self.configuration.alwaysOpenSafari
+        && [SFSafariViewController class] != nil) {
+        SFSafariViewController *safariVC = [[SFSafariViewController alloc] initWithURL:redirectURL];
+        safariVC.delegate = self;
+        self.presentedViewController = safariVC;
+        [presentingViewController presentViewController:safariVC
+                                               animated:YES
+                                             completion:nil];
+    }
+    else {
+        // TODO: call openURL
+        [[UIApplication sharedApplication] openURL:redirectURL];
+    }
+
+    return YES;
+}
+
+- (BOOL)redirectAllowedForSource:(STPSource *)source {
+    // TODO: verify that they are valid URLs and that return url is set up correctly
+    return ((source.flow == STPSourceFlowRedirect)
+            && source.redirect[@"url"] != nil
+            && source.redirect[@"return_url"] != nil);
+}
+
+- (void)cleanupAndCompleteRedirectWithError:(nullable NSError *)error {
+    STPSource *source = self.inProgressAuthSource;
+
+    [[STPURLCallbackHandler shared] unregisterListener:self
+                                                forURL:[NSURL URLWithString:source.redirect[@"return_url"]]];
+
+
+    if (self.presentedViewController) {
+        [self.presentedViewController dismissViewControllerAnimated:YES
+                                                         completion:^{
+                                                             if (self.completion) {
+                                                                 self.completion(source, error);
+                                                             }
+                                                             self.presentedViewController = nil;
+                                                         }];
+    }
+    else {
+        if (self.completion) {
+            self.completion(source, error);
+        }
+    }
+
+    _inProgressAuthSource = nil;
+    self.completion = nil;
+
+}
+
+- (BOOL)handleURLCallback:(NSURL *)url {
+    if (url == nil
+        || self.inProgressAuthSource == nil) {
+        [self cleanupAndCompleteRedirectWithError:nil];
+        return NO;
+    }
+    else {
+
+        NSURLComponents *components = [[NSURLComponents alloc] initWithURL:url
+                                                   resolvingAgainstBaseURL:NO];
+
+        NSMutableDictionary <NSString *, NSString *> *queryItems = [NSMutableDictionary new];
+        for (NSURLQueryItem *queryItem in components.queryItems) {
+            queryItems[queryItem.name] = queryItem.value;
+        }
+
+        NSString *clientSecretString = queryItems[@"client_secret"];
+//        NSString *livemodeString = queryItems[@"livemode"];
+        NSString *sourceIdString = queryItems[@"source"];
```

Comment:
it'd be nice to extract this query item parsing stuff to a category (e.g. on `NSURL`) and test it